src/core_init_atmosphere/mpas_geotile_manager.F

```diff
@@ -0,0 +1,1025 @@
+        c_tile_ptr = c_loc(tile % tile)
+        call read_geogrid(c_fname, c_tile_ptr, c_tile_nx, c_tile_ny, c_tile_nz, c_signed, c_endian, &
+                            c_scalefactor, c_wordsize, status)
+        if (status /= 0) then
+            call mpas_log_write("Error reading this geogrid file: "//trim(fname), messageType=MPAS_LOG_ERR)
+            ierr = 1
+            return
+        endif
+
+        tile % fname = fname
+        tile % hash_x = (start_x - 1) / tile_nx
+        tile % hash_y = (start_y - 1) / tile_ny
+        tile % x = start_x
+        tile % y = start_y
+
+        !
+        ! Add the tile to the hash table
+        !
+        mgr % hash(tile % hash_x, tile % hash_y)  % ptr => tile
+
+    end function mpas_geotile_mgr_add_tile
+
+
+    !***********************************************************************
+    !
+    !  private function mpas_geotile_mgr_gen_tile_name => gen_filename
+    !
+    !> \brief   Generate the filename of the tile at lat, lon
+    !> \author  Miles A. Curry
+    !> \date    02/20/2020
+    !> \details
+    !> Generate the name of the file that contains the coordinate lat, lon
+    !> (in radians) and optionally return start_x and start_y (the location
+    !> of the first global pixel coordinate of tile). Warning: This function
+    !> can produce filenames that may not exist (For lon less than -.5 * pi and
+    !> greater than .5 * pi and lat less than -pi and greater than pi).
+    !
+    !-----------------------------------------------------------------------
+    function mpas_geotile_mgr_gen_tile_name(mgr, lat, lon, fname, start_x, start_y) result(ierr)
+
+        implicit none
+
+        class (mpas_geotile_mgr_type) :: mgr
+        real (kind=RKIND), value :: lat
+        real (kind=RKIND), value :: lon
+        character (len=StrKIND), intent(out) :: fname
+        integer, intent(out), optional :: start_x
+        integer, intent(out), optional :: start_y
+
+        character (len=StrKIND), parameter :: fname_format = "(i5.5, '-', i5.5, '.', i5.5, '-', i5.5)"
+
+        real (kind=RKIND), pointer :: dx
+        real (kind=RKIND), pointer :: dy
+        integer, pointer :: tile_nx
+        integer, pointer :: tile_ny
+        integer, dimension(2) :: x
+        integer, dimension(2) :: y
+
+        integer :: ierr
+        ierr = 0
+
+        call mpas_pool_get_config(mgr % pool, 'tile_x', tile_nx)
+        call mpas_pool_get_config(mgr % pool, 'tile_y', tile_ny)
+        call mpas_pool_get_config(mgr % pool, 'dx', dx)
+        call mpas_pool_get_config(mgr % pool, 'dy', dy)
+
+        ! Find the global pixel location that contains lat, lon
+        call mgr % latlon_to_pixel(lat, lon, x(1), y(1))
+
+        ! Calculate the range of this tile, which will be its filename
+        x(1) = (x(1) - modulo(x(1), tile_nx)) + 1
+        x(2) = x(1) + tile_nx
+        x(2) = x(2) - modulo(x(2), tile_nx)
```

Comment:
I may be overlooking some special cases, but this logic seems more complicated than necessary. We know that `x(1)` is one more than a multiple of `tile_nx`, and by line 584, so is `x(2)`; so, couldn't we just subtract 1 from `x(2)`?